### PR TITLE
CSF3: Generate default titles based on file path

### PR DESCRIPTION
**Pull request by shilman**
Issue: #11186 

## What I did

Generate default titles based on file path. With this PR the `stories` field in `.storybook/main.js` can take on three possible values:

```js
module.exports = {
  stories: = [
    '../packages/foo/**/*.stories.mdx', // glob
    { directory: '../packages/baz', titlePrefix: 'Baz Components', files: '*.stories.*' }, // specifier
    '../packages/bar', // directory
  ]
}
```

- A file glob that matches story files => same as CSF2
- A specifier object containing:
  - A directory that contains story files somewhere beneath to
  - The "titlePrefix" that the directory maps to (optional, defaults to `''`)
  - A file selector glob (optional, defaults to `'*.stories.@(mdx|tsx|ts|jsx|js)'`)
- A path to a directory which is equivalent to the specifier object

If you use either of the second two objects and have the `features.previewCsfV3` flag set, you no longer need to fill in the `default.title` annotation in CSF.

## What I need

~~This currently fails in static builds of Storybook because the `fileName` is rewritten as a number rather than the actual file path.~~

self-merging @ndelangen @tmeasday with the rename of `root` to `titlePrefix` per Norbert's feedback

## How to test

- [ ] See `examples/react-ts/main.ts` / `AccountForm.stories.tsx`
- [ ] See attached tests

### CODE REVIEW

File: lib/core-client/src/preview/autoTitle.ts [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+  if (fileName.startsWith(directory)) {
+    const suffix = fileName.replace(directory, '');
+    return stripExtension(path.join(root, suffix));
+  }
+  return undefined;
+};
+
+const autoTitleV3 = (meta: Meta, fileName: string) => {
+  if (meta.title) return meta.title;
+  console.log({ fileName, STORIES });
```
Comment (tmeasday) [inline lib/core-client/src/preview/autoTitle.ts:39]:
```suggestion
```

File: lib/core-client/src/preview/autoTitle.test.ts [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+describe('autoTitle', () => {
+  it('no directory', () => {
+    expect(auto('/path/to/file', { glob: '' })).toBeFalsy();
+  });
+
+  it('no match', () => {
+    expect(auto('/path/to/file', { glob: '', specifier: { directory: '/other' } })).toBeFalsy();
+  });
+
+  describe('trailing slash', () => {
```
Comment (tmeasday) [inline lib/core-client/src/preview/autoTitle.test.ts:17]: Aren't these describe names backwards?

File: lib/core-common/src/utils/__tests__/normalize-stories.test.ts [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+        "glob": "../**/*.stories.mdx",
+        "specifier": {
+          "directory": "..",
+          "root": "",
+          "files": "*.stories.mdx"
+        }
+      }
+    `);
+  });
+  it('directory/root specifier', () => {
```
Comment (tmeasday) [inline lib/core-common/src/utils/__tests__/normalize-stories.test.ts:63]: Add a test for all three?

File: lib/core-common/src/utils/__tests__/normalize-stories.test.ts [hunk 4]
```diff
@@ -0,0 +1,75 @@
+import { normalizeStoriesEntry, normalizeStories } from '../normalize-stories';
+
+expect.addSnapshotSerializer({
+  print: (val: any) => JSON.stringify(val, null, 2),
+  test: (val) => typeof val !== 'string',
+});
```
Comment (tmeasday) [inline lib/core-common/src/utils/__tests__/normalize-stories.test.ts:6]: TBH I don't really understand why you are using snapshots here, would it not be easier to just use `.toMatchObject()`? I guess it's your call.

**Review by tmeasday — COMMENTED** [review]
So this look good, but I guess the static build thing is a bit of a deal breaker.

Similarly, non webpack builders might have problems with a feature that relies on `parameters.fileName`.

To me, it feels like this is something that will work well when we are generating `stories.json`.

File: lib/core-common/src/utils/__tests__/normalize-stories.test.ts (hunk 4 above)
Comment (shilman) [reply lib/core-common/src/utils/__tests__/normalize-stories.test.ts:6]: IMO snapshot tests are easier to maintain

**Comment by bebraw** [comment]
> This currently fails in static builds of Storybook because the fileName is rewritten as a number rather than the actual file path.

What's the easiest way to reproduce this behavior?

It sounds like there's some mismatch between dev/prod configurations assuming it's a webpack problem. Looking at the code, it's around line `const title = autoTitle(defaultExport, fileName)` where I would drop a `debugger;` statement to see what's going on.

**Comment by shilman** [comment]
@bebraw sorry I missed your comment. To reproduce the problem:

```
cd examples/react-ts
yarn storybook # <-- works
yarn build-storybook # <-- fails
```

**Comment by bebraw** [comment]
I can reproduce the issue now and I have a rough idea how you might be able to fix it. Here are a couple of screenshots to illustrate the problem.

**yarn storybook*

Screen Shot 2021-06-29 at 12 18 48

**yarn build-storybook**

Screen Shot 2021-06-29 at 12 18 42

`fileName` expected by `autoTitleFromEntry` isn't what you expect and during a build it seems to refer to a numeric module id.

I think one way to fix this would be to change webpack configuration in the latter case to include `optimization.moduleIds = 'named'`. When webpack runs in production mode, it uses numbers for module ids and going with `named` would change the behavior here as well so you get what you expect.

The side effect of doing this is that it will increase the bundle size slightly since names are longer than numbers but it's possible this doesn't make a big difference.

**Comment by shilman** [comment]
Thanks for the tip @bebraw -- works great!!! 🙏

**Comment by marcus13371337** [comment]
Hello @shilman ,

Great with this new feature, will make a lot of things cleaner in our monorepo setup. ❤️

However, I can't see the `titlePrefix`  I'm providing anywhere in the UI. What am I doing wrong? Tried with `--no-manager-cache`

**Comment by marcus13371337** [comment]
Found the issue, apparently, you can't specify a manual title if you use the `titlePrefix`? Just reading the parameter name, I would expected that what I'm passing to the `titlePrefix` would be the prefix to any title no matter if I specified manually or if it was auto generated from the path

**Comment by shilman** [comment]
@marcus13371337 If you specify the title, it uses that title verbatim. We'll document all this as part of of the 6.4 release (cc @jonniebigodes)

**Comment by spacecakes** [comment]
Love this feature as it has allowed me to get rid of hundreds of lines of code. Just wondering if there's a way to get Storybook to flatten the sidebar folder structure when the directory name is the same as the story file name, so we can avoid having this redundant nesting: 

Screenshot 2022-04-22 at 14 39 11

Is there any way for me to configure that?

**Comment by shilman** [comment]
@spacecakes  fixed in 6.5


Try upgrading to the latest prerelease:

```
npx sb@next upgrade --prerelease
```

**Comment by spacecakes** [comment]
> @spacecakes fixed in 6.5
> 
> Try upgrading to the latest prerelease:
> 
> ```
> npx sb@next upgrade --prerelease
> ```

Wow, that is convenient. Thank you very much!

**Comment by spacecakes** [comment]
> @spacecakes fixed in 6.5

Wow, that is convenient. Thank you very much!

The only issue I noticed with this is that story hoisting does not work for multi-word component names, because, say, `CheckboxGroup.stories.js` is gets processed into `Checkbox Group` for the story name, while the path remains as-is so they no longer match. So we get this:

Screenshot 2022-04-22 at 18 58 16

Not a dealbreaker by any means and I realise it's prelease software, but it would be nice if this worked to avoid nesting single-story components with this setup.

**Pull request merged**